aws-android-sdk-mobile-client/src/androidTest/java/com/amazonaws/mobile/client/AWSMobileClientCustomAuthTest.java

```diff
@@ -0,0 +1,172 @@
+/*
+ * Copyright 2019 Amazon.com, Inc. or its affiliates.
+ * All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.amazonaws.mobile.client;
+
+import android.content.Context;
+import android.support.test.InstrumentationRegistry;
+import android.util.Log;
+
+
+import com.amazonaws.auth.AWSCredentials;
+import com.amazonaws.mobile.client.results.SignInResult;
+import com.amazonaws.mobile.client.results.Token;
+import com.amazonaws.mobile.client.results.Tokens;
+import com.amazonaws.mobile.config.AWSConfiguration;
+import com.amazonaws.mobileconnectors.cognitoidentityprovider.util.CognitoServiceConstants;
+import org.junit.BeforeClass;
+import org.junit.Test;
+
+import java.util.Date;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.concurrent.CountDownLatch;
+
+import static com.amazonaws.mobile.auth.core.internal.util.ThreadUtils.runOnUiThread;
+import static org.junit.Assert.assertNotNull;
+import static org.junit.Assert.assertTrue;
+
+public class AWSMobileClientCustomAuthTest extends AWSMobileClientTestBase {
+
+    private static Context appContext;
+    private static AWSMobileClient auth;
+
+    private static final CountDownLatch signUpLatch = new CountDownLatch(1);
+    private static final String USERNAME = "roskumr@amazon.com";
+    private static final String PASSWORD = "Test@123";
+    private static final String TAG = AWSMobileClientCustomAuthTest.class.getSimpleName();
+
+    @BeforeClass
+    public static void beforeClass() throws Exception {
+        appContext = InstrumentationRegistry.getTargetContext();
+        final CountDownLatch latch = new CountDownLatch(1);
+        final AWSConfiguration awsConfiguration = new AWSConfiguration(appContext);
+        awsConfiguration.setConfiguration("Auth1");
+        assert awsConfiguration.optJsonObject("Auth").has("authenticationFlowType");
+        AWSMobileClient.getInstance().initialize(appContext, awsConfiguration, new Callback<UserStateDetails>() {
+            @Override
+            public void onResult(UserStateDetails result) {
+                latch.countDown();
+            }
+
+            @Override
+            public void onError(Exception e) {
+                Log.e(TAG, "onError: ", e);
+                latch.countDown();
+            }
+        });
+        latch.await();
+        auth = AWSMobileClient.getInstance();
+    }
+
+    @Test
+    public void testCustomAuth() throws Exception {
```

Comment:
Can you add a README or documentation for this test about the backend setups that need to be in place for this integration test to succeed? I assume that we need some kind of Lambda in place to send & retrieve the auth challenge?

Reply:
Added